packages/xo-web/src/xo-app/disk-import/index.js

```diff
@@ -0,0 +1,190 @@
+import Dropzone from 'dropzone'
+import fromEvent from 'promise-toolbox/fromEvent'
+import React, { Component } from 'react'
+import { Container } from 'grid'
+import { formatSize, mapPlus } from 'utils'
+import { generateId, linkState } from 'reaclette-utils'
+import { importDisks } from 'xo'
+import { injectIntl } from 'react-intl'
+import { injectState, provideState } from 'reaclette'
+import { InputCol, LabelCol, Row } from 'form-grid'
+import { readCapacityAndGrainTable } from 'xo-vmdk-to-vhd'
+import { SelectSr } from 'select-objects'
+
+const getInitialState = () => ({
+  disks: [],
+  mapDescriptions: {},
+  mapNames: {},
+  sr: undefined,
+})
+
+const DiskImport = decorate([
+  provideState({
+    initialState: getInitialState,
+    effects: {
+      handleDrop: (effects, files) => async ({ sr }) => {
+        const disks = await Promise.all(
+          mapPlus(files, async (file, push) => {
+            const { name } = file
+            const extIndex = name.lastIndexOf('.')
+            let type
+            if (
+              extIndex >= 0 &&
+              (type = name.slice(extIndex + 1)) &&
+              (type === 'vmdk' || type === 'vhd')
+            ) {
+              let vmdkData
+              if (type === 'vmdk') {
+                const parsed = await effects.parseVmdk(file)
+                vmdkData = {
+                  blocksTable: await parsed.tablePromise,
+                  capacity: parsed.capacityBytes,
+                }
+              }
+              push({
+                id: generateId(),
+                file,
+                name,
+                sr,
+                type,
+                vmdkData,
+              })
+            }
+          })
+        )
+        return { disks }
+      },
+      linkState,
+      onChangeDescription: (_, { target: { name, value } }) => ({
+        mapDescriptions,
+      }) => {
+        mapDescriptions[name] = value
+        return { mapDescriptions }
+      },
+      onChangeName: (_, { target: { name, value } }) => ({ mapNames }) => {
+        mapNames[name] = value
+        return { mapNames }
+      },
+      onChangeSr: (_, sr) => ({ sr }),
+      parseVmdk: async file =>
+        readCapacityAndGrainTable(async (start, end) => {
+          /* global FileReader */
+          const reader = new FileReader()
+          reader.readAsArrayBuffer(file.slice(start, end))
+          return (await fromEvent(reader, 'loadend')).target.result
+        }),
+      reset: getInitialState,
+    },
+  }),
+  injectIntl,
+  injectState,
+  class extends Component {
+    _import = () => {
+      const {
+        state: { disks, mapDescriptions, mapNames, sr },
+      } = this.props
+      return importDisks(
+        disks.map(disk => {
+          const { id, name } = disk
+          return {
+            ...disk,
+            name: mapNames[id] || name,
+            description: mapDescriptions[id],
+          }
+        }),
+        sr
+      )
+    }
```

Comment:
Why isn't this an effect?

Reply:
for `redirectOnSuccess` to work.

Reply:
Right! Then please add a comment because I pretty sure I'll forget it again :)

Reply:
Also, `redirectOnSuccess` doesn't seem to need the returned value in this case.

Reply:
but it needs to wait `handler` to be resolved.

Reply:
But an effect returns a promise that settles only when the promise *you* returned settles.
@julien-f can you confirm that?
```js
effects: {
  async foo() {
    await Promise.resolve()
    console.log('foo')
  }
}
// ...
({ effects }) => {
  effects.foo().then(() => console.log('bar'))
  return null
}
```
should work and print `foo` then `bar`, right?

Reply:
Yep.